internal/controllers/topology/cluster/desired_state.go

```diff
@@ -797,65 +805,49 @@ func computeMachineDeploymentVersion(s *scope.Scope, machineDeploymentTopology c
 	// NOTE: this case should never happen (upgrading a MachineDeployment) before creating a CP,
 	// but we are implementing this check for extra safety.
 	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
-		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(desiredMDName)
+		return currentVersion
 	}
 
 	// If the current control plane is upgrading, then do not pick up the desiredVersion yet.
 	// Return the current version of the machine deployment. We will pick up the new version after the control
 	// plane is stable.
-	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to check if control plane is upgrading")
-	}
-	if cpUpgrading {
-		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-		return currentVersion, nil
+	if s.UpgradeTracker.ControlPlane.IsUpgrading {
+		s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(desiredMDName)
+		return currentVersion
 	}
 
 	// If control plane supports replicas, check if the control plane is in the middle of a scale operation.
 	// If the current control plane is scaling, then do not pick up the desiredVersion yet.
 	// Return the current version of the machine deployment. We will pick up the new version after the control
 	// plane is stable.
 	if s.Blueprint.Topology.ControlPlane.Replicas != nil {
-		cpScaling, err := contract.ControlPlane().IsScaling(s.Current.ControlPlane.Object)
-		if err != nil {
-			return "", errors.Wrap(err, "failed to check if the control plane is scaling")
-		}
-		if cpScaling {
-			s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(currentMDState.Object.Name)
-			return currentVersion, nil
+		if s.UpgradeTracker.ControlPlane.IsScaling {
+			s.UpgradeTracker.MachineDeployments.MarkPendingUpgrade(desiredMDName)
+			return currentVersion
 		}
 	}
 
 	// Check if we are about to upgrade the control plane. In that case, do not upgrade the machine deployment yet.
 	// Wait for the new upgrade operation on the control plane to finish before picking up the new version for the
 	// machine deployment.
-	currentCPVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to get version of current control plane")
-	}
-	desiredCPVersion, err := contract.ControlPlane().Version().Get(desiredControlPlaneState.Object)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to get version of desired control plane")
-	}
-	if *currentCPVersion != *desiredCPVersion {
+	if s.UpgradeTracker.ControlPlane.IsStartingUpgrade {
```

Comment:
I think it would be a good simplification to move all those cp checks into a util func like isControlPlaneStable